src/main/scala/org/broadinstitute/dsde/firecloud/EntityService.scala

```diff
@@ -305,41 +306,62 @@ class EntityService(rawlsDAO: RawlsDAO, importServiceDAO: ImportServiceDAO, goog
 
         val rand = java.util.UUID.randomUUID.toString.take(8)
         val bagItFile = File.createTempFile(s"$rand-samples", ".tsv")
+        var bytesDownloaded = new AtomicLong(-1) // careful, this is a var
 
         try {
           val conn = bagitURL.openConnection()
           val length = conn.getContentLength
           conn.asInstanceOf[HttpURLConnection].disconnect()
 
-          if (length > Rawls.entityBagitMaximumSize) {
+          if (length == 0) {
+            Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, s"BDBag has content-length 0"))
+          } else if (length > Rawls.entityBagitMaximumSize) {
             Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, s"BDBag size is too large."))
           } else {
-            //this magic creates a process that downloads a URL to a file (which is #>), and then runs the process (which is !!)
-            bagitURL #> bagItFile !!
-
-            //make two big strings containing the participants and samples TSVs
-            //if i could turn back time this would use streams to save memory, but hopefully this will all go away when entity service comes along
-            unzipTSVs(bagitRq.bagitURL, new ZipFile(bagItFile.getAbsolutePath)) { (participantsStr, samplesStr) =>
-              (participantsStr, samplesStr) match {
-                case (None, None) =>
-                  Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "You must have either (or both) participants.tsv and samples.tsv in the zip file"))
-                case _ =>
-                  for {
-                    // This should vomit back errors from rawls.
-                    participantResult <- participantsStr.map(ps => importEntitiesFromTSV(workspaceNamespace, workspaceName, ps, userInfo)).getOrElse(Future.successful(RequestComplete(OK)))
-                    sampleResult <- samplesStr.map(ss => importEntitiesFromTSV(workspaceNamespace, workspaceName, ss, userInfo)).getOrElse(Future.successful(RequestComplete(OK)))
-                  } yield {
-                    participantResult match {
-                      case RequestComplete((OK, _)) => sampleResult
-                      case _ => participantResult
+            // download the file
+            val readFromBagit = Channels.newChannel(bagitURL.openStream())
```

Comment:
I'm not too familiar with `java.nio.channels.Channels`, but I'm assuming it's generally a better approach than the one being replaced?

Reply:
it's definitely less mysterious, and offers more insight, such as the number of bytes downloaded